Classes/System/Dispatcher.php

```diff
@@ -26,31 +26,82 @@
  ***************************************************************/
 
 use Aoe\Restler\System\Restler\Builder as RestlerBuilder;
+use Psr\Http\Message\ResponseInterface;
+use Psr\Http\Message\ServerRequestInterface;
+use Psr\Http\Server\MiddlewareInterface;
+use Psr\Http\Server\RequestHandlerInterface;
+use TYPO3\CMS\Core\Http\Response;
+use TYPO3\CMS\Core\Http\Stream;
+use TYPO3\CMS\Core\Utility\GeneralUtility;
+use TYPO3\CMS\Extbase\Object\ObjectManager;
 
 /**
  * @package Restler
  */
-class Dispatcher
+class Dispatcher implements MiddlewareInterface
 {
     /**
      * @var RestlerBuilder
      */
     private $restlerBuilder;
 
-    /**
-     * @param RestlerBuilder $restlerBuilder
-     */
-    public function __construct(RestlerBuilder $restlerBuilder)
+    public function __construct(ObjectManager $objectManager = null)
     {
-        $this->restlerBuilder = $restlerBuilder;
+        if (!$objectManager) {
+            $objectManager = GeneralUtility::makeInstance(ObjectManager::class);
+        }
+        $this->restlerBuilder = $objectManager->get(RestlerBuilder::class);
     }
 
     /**
-     * dispatch the REST-API-request
+     * Process an incoming server request.
+     *
+     * Processes an incoming server request in order to produce a response.
+     * If unable to produce the response itself, it may delegate to the provided
+     * request handler to do so.
      */
-    public function dispatch()
+    public function process(ServerRequestInterface $request, RequestHandlerInterface $handler): ResponseInterface
+    {
+        $restlerObj = $this->restlerBuilder->build($request);
+
+        if ($this->isRestlerUrl('/' . $restlerObj->url)) {
```

Comment:
This approach adds an overhead of initializing restler to ALL requests reaching TYPO3.
v8 version was more lightweight, because it first checked for occurence of "/api/" before initializating restler framework and doing more checks.
I would prefer to reintroduce this approach, and move initialization from constructor to after positive check that url contains /api/ prefix.

Reply:
Best way would be to fix the Extbase code to check for an initialized TemplateService before using it or to switch the typoscript config loading to use config instead of tmpl->setup.
But in the meantime I'll move the initialization further down the callstack.

Reply:
After a quick check with emptying out the constructor and replacing the body of the process method with the pass through `return $handler->handle($request);` there is no discernable perfomance difference.

Could you run your setup through blackfire.io for profiling?

Reply:
> Best way would be to fix the Extbase code to check for an initialized TemplateService before using it or to switch the typoscript config loading to use config instead of tmpl->setup.

Can you submitt an issue or even a patch for the core? I can take it further and discuss with Alex S who is maintaining Extbase code base (ad doing a lot of great changes in that area).

Reply:
> After a quick check with emptying out the constructor and replacing the body of the process method with the pass through `return $handler->handle($request);` there is no discernable perfomance difference.
> 
> Could you run your setup through blackfire.io for profiling?

I did blackfire profiling. Will send you trace on slack.